aggregator/elem.go

```diff
@@ -447,17 +534,26 @@ func (e *GaugeElem) Consume(earlierThanNanos int64, fn aggMetricFn) bool {
 		if e.values[idx].timeNanos >= earlierThanNanos {
 			break
 		}
-		endAtNanos := e.values[idx].timeNanos + int64(e.policy.Resolution().Window)
-		e.processValue(endAtNanos, e.values[idx].gauge, fn)
 		idx++
 	}
+	e.toConsume = e.toConsume[:0]
 	if idx > 0 {
 		// Shift remaining values to the left and shrink the values slice
+		e.toConsume = append(e.toConsume, e.values[:idx]...)
 		n := copy(e.values[0:], e.values[idx:])
 		e.values = e.values[:n]
 	}
 	canCollect := len(e.values) == 0 && e.tombstoned
 	e.Unlock()
+
+	for i := range e.toConsume {
+		endAtNanos := e.toConsume[i].timeNanos + int64(e.policy.Resolution().Window)
+		e.toConsume[i].gauge.Lock()
```

Comment:
Same comment as for counters, is it possible to use an atomic variable here? Gauges are a little trickier since they're floats but we could store the value as a uint64 and then convert to and from a float when we get and set its value. I'm not familiar with the requirements of the aggregation function `fn` so it may not be viable, figured I would suggest it just in case though.